### ISSUE TITLE
[Flang] Incorrect diagnostic on generic type-bound procedure expansion.

### ISSUE THREAD

**Issue by DanielCChen** [llvm/llvm-project#124777]
Consider the following code:
```
module m

   type base
      integer i
      contains
         procedure, pass :: add3
         procedure :: adde
         generic :: operator(+) => add3
   end type

   interface operator(+)
      module procedure adde
   end interface

   contains

   type(base) elemental function adde (a,b)
      class(base), intent(in) :: a, b
   end function

   type(base) function add3 (a,b)
      class(base), intent(in) :: a, b(:,:,:)
   end function

end module

program genericOperatorResolve007a
   use m

   type(base) :: b0
   type(base), allocatable :: b3(:,:,:)

   b0 = b0 + b3                                     !! ERROR not expected

end program
```

Flang currently issues an error as:
```
error: Semantic errors in t.f
./t.f:33:4: error: No intrinsic or user-defined ASSIGNMENT(=) matches scalar TYPE(base) and rank 3 array of TYPE(base)
     b0 = b0 + b3
     ^^^^^^^^^^^^
```

It seems the generic interface expansion of `adde` is the culprit as the code passed after I removed it.
All ifort , gfortran and XLF compile the code successfully.

**Comment by klausler** [comment]
The program is ambiguous; both the type-bound and non-type-bound generic interfaces for `OPERATOR(+)` can apply to the expression `b0 + b3`.

**Comment by DanielCChen** [comment]
> The program is ambiguous; both the type-bound and non-type-bound generic interfaces for `OPERATOR(+)` can apply to the expression `b0 + b3`.

The interface of the type-bound procedure and the. non-type-bound generic interface is different in this code (otherwise, it would be ambiguous). 
`b0` is scalar and `b3` is rank-3. The operation `+` should be resolved to `add3` as it has matching characteristics.

**Comment by klausler** [comment]
Right, non-elemental specific procedures take precedence over elemental specific procedures.  There's code in Semantics/expression.cpp to implement this distinction; perhaps it got broken.  Will fix.
